Sources/MongoSwift/MongoDatabase.swift

```diff
@@ -223,30 +212,64 @@ public class MongoDatabase {
      *   - options: Optional `ListCollectionsOptions` to use when executing this command
      *   - session: Optional `ClientSession` to use when executing this command
      *
-     * - Returns: a `MongoCursor` over an array of collections
+     * - Returns: a `MongoCursor` over an array of `CollectionSpecification`s
      *
      * - Throws:
      *   - `userError.invalidArgumentError` if the options passed are an invalid combination.
      *   - `UserError.logicError` if the provided session is inactive.
      */
     public func listCollections(_ filter: Document? = nil,
                                 options: ListCollectionsOptions? = nil,
-                                session: ClientSession? = nil) throws -> MongoCursor<Document> {
-        var opts = try encodeOptions(options: options, session: session)
-        if let filterDoc = filter {
-            opts = opts ?? Document()
-            // swiftlint:disable:next force_unwrapping
-            opts!["filter"] = filterDoc // guaranteed safe because of nil coalescing default.
+                                session: ClientSession? = nil) throws -> MongoCursor<CollectionSpecification> {
+        let operation = ListCollectionsOperation(database: self, filter: filter, options: options, nameOnly: false)
+        guard case let .specs(result) = try self._client.executeOperation(operation, session: session) else {
+            throw RuntimeError.internalError(message: "Invalid result")
         }
+        return result
+    }
 
-        return try MongoCursor(client: self._client, decoder: self.decoder, session: session) { conn in
-            self.withMongocDatabase(from: conn) { dbPtr in
-                guard let collections = mongoc_database_find_collections_with_opts(dbPtr, opts?._bson) else {
-                    fatalError(failedToRetrieveCursorMessage)
-                }
-                return collections
-            }
+    /**
+     * Gets a list of `MongoCollection`s in this database.
+     *
+     * - Parameters:
+     *   - filter: a `Document`, optional criteria to filter results by
+     *   - options: Optional `ListCollectionsOptions` to use when executing this command
+     *   - session: Optional `ClientSession` to use when executing this command
+     *
+     * - Returns: An array of `MongoCollection`s that match the provided filter.
+     *
+     * - Throws:
+     *   - `userError.invalidArgumentError` if the options passed are an invalid combination.
+     *   - `UserError.logicError` if the provided session is inactive.
+     */
+    public func listMongoCollections(_ filter: Document? = nil,
+                                     options: ListCollectionsOptions? = nil,
+                                     session: ClientSession? = nil) throws -> [MongoCollection<Document>] {
```

Comment:
hmm, maybe, though unless if all your collections are structured the same way it might not be that useful. 
we could maybe add another variant that takes in a type parameter but I think we should leave this one as it is, for symmetry with the `collection` methods on DB.